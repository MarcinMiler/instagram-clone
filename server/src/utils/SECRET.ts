export default `skjnlad9f823l8ay09h23r`
