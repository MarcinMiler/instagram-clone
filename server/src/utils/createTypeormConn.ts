import { getConnectionOptions, createConnection } from 'typeorm'

export const createTypeormConn = async () => {
    const connectionOptions = await getConnectionOptions(process.env.NODE_ENV)
    return createConnection({ ...connectionOptions, name: 'default' })
}
