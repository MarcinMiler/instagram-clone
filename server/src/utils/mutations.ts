export const registerMutation = (
    email: string,
    password: string,
    fullname: string,
    username: string
) => `
    mutation {
        register(email: "${email}", password: "${password}", fullname: "${fullname}", username: "${username}") {
            ok,
            error
        }
    }
`

export const loginMutation = (email: string, password: string) => `
    mutation {
        login(email: "${email}", password: "${password}") {
            ok,
            error
        }
    }
`

export const loginMutationWithToken = (email: string, password: string) => `
    mutation {
        login(email: "${email}", password: "${password}") {
            token
        }
    }
`

export const addPhoto = (url: string, text: string) => `
    mutation {
        addPhoto(url: "${url}", text: "${text}")
    }
`

export const likePhoto = (photoId: string) => `
    mutation {
        likePhoto(photoId: "${photoId}")
    }
`
