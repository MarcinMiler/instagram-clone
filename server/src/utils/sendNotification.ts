import { Notification } from '../entity/Notification'
import { User } from '../entity/User'

const typeOfMessage = (type: string) => {
    interface Types {
        [key: string]: string
    }

    const types: Types = {
        comment: 'commented your photo.',
        like: 'liked your photo.',
        likeComment: 'liked your comment',
        follow: 'followed you.'
    }
    return types[type]
}

export const sendNotification = async (
    currentUser: number,
    userToSend: number,
    messageType: string,
    expo: any,
    photoId?: number
) => {
    const message = typeOfMessage(messageType)

    const notification = Notification.create({
        userId: currentUser,
        photoId,
        message,
        sendTo: userToSend
    })
    await notification.save()

    const sendTo: any = await User.findOneById(userToSend)
    const u: any = await User.findOneById(currentUser)

    if (sendTo.notificationToken) {
        const expoMessage = {
            to: sendTo.notificationToken,
            sound: 'default',
            body: `${u.username} ${message}`
        }

        await expo.sendPushNotificationsAsync([expoMessage])
    }
}
