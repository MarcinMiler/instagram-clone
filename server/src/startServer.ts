import 'reflect-metadata'
import { importSchema } from 'graphql-import'
import { GraphQLServer } from 'graphql-yoga'
import * as path from 'path'
import * as fs from 'fs'
import * as jwt from 'jsonwebtoken'
import { mergeSchemas, makeExecutableSchema } from 'graphql-tools'
import { GraphQLSchema } from 'graphql'
import { createTypeormConn } from './utils/createTypeormConn'
import SECRET from './utils/SECRET'

const Expo = require('expo-server-sdk')

const options = {
    port: process.env.NODE_ENV === 'test' ? 0 : 4000,
    endpoint: '/graphql',
    cors: {
        origin: '*'
    }
}

const addUser = (req: any, _: any, next: any) => {
    const token = req.headers.token
    if (token) {
        const user = jwt.verify(token, SECRET)
        req.user = user
    }
    next()
}

export const startServer = async () => {
    const schemas: GraphQLSchema[] = []
    const folders = fs.readdirSync(path.join(__dirname, './modules'))
    folders.forEach(folder => {
        const { resolvers } = require(`./modules/${folder}/resolvers`)
        const typeDefs = importSchema(
            path.join(__dirname, `./modules/${folder}/schema.graphql`)
        )
        schemas.push(makeExecutableSchema({ resolvers, typeDefs }))
    })

    const expo = new Expo()

    const server = new GraphQLServer({
        schema: mergeSchemas({ schemas }),
        context: (req: any) => {
            if (req.request.user) {
                return {
                    user: req.request.user.id,
                    expo
                }
            }
            return
        }
    })
    server.use(addUser)

    await createTypeormConn()

    const app = await server.start(options)

    return app
}
