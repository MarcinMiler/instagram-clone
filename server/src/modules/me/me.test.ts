import { request, GraphQLClient } from 'graphql-request'
import { startServer } from '../../startServer'
import { loginMutationWithToken, registerMutation } from '../../utils/mutations'

let getHost = () => ''
let app: any

beforeAll(async () => {
    app = await startServer()
    const { port } = app.address()
    getHost = () => `http://127.0.0.1:${port}`
})

const email = 'm@m.com'
const password = 'mm'
const fullname = 'Marcin Miler'
const username = 'Marcinek'

const query = `
    {
        me {
            id,
            fullname,
            username
        }
    }
`

describe('Query me', async () => {
    it('should return user', async () => {
        await request(
            getHost(),
            registerMutation(email, password, fullname, username)
        )
        const login: any = await request(
            getHost(),
            loginMutationWithToken(email, password)
        )
        const { token } = login.login

        const client = new GraphQLClient(getHost(), {
            headers: {
                token
            }
        })

        const response = await client.request(query)

        expect(response).toMatchSnapshot()
    })
})
