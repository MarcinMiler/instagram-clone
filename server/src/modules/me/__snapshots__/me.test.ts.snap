// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query me should return user 1`] = `
Object {
  "me": Object {
    "fullname": "Marcin Miler",
    "id": "1",
    "username": "Marcinek",
  },
}
`;
